api_service: Rework cgroup detection

euank opened the pull request:
Use the `subcgroup` file hint provided by some stage1s rather than
machined registration.

Fixes #3060

TODO still:

- [x] Validate this use of `subcgroup` makes sense / is correct.
- [x] Minimal validation this works with stage1-kvm.
- [ ] Minimum validation of cadvisor's consuming of this API.
- [x] Better tests? (though `TestAPIServiceCgroup` exists and passes)
- [ ] In the future move the contract to a `cgroup` file instead that all stage1's write perhaps.

cc @yifan-gu @alban @iaguis for opinions on if this implementaiton makes sense

yifan-gu commented on rkt/api_service.go:
```diff
@@ -333,6 +332,43 @@ func (s *v1AlphaAPIServer) getPodManifest(p *pod) (*schema.PodManifest, []byte,
+	}
+	// if it is an "isNotExist", assume this is a stage1 flavor that won't change
+	// cgroups. Just give the systemd cgroup of its pid
+	cgroup, err := cgroup.GetCgroupPathByPid(pid, "name=systemd")
```
Isn't this assume the host have systemd?

euank replied:
This code was just moved from its previous location, so that assumption already existed.

yifan-gu replied:
@euank OK, can we add a comment to mention this?
Seems this return error on non-systemd distros, cc @alban @lucab @iaguis

yifan-gu commented on rkt/api_service.go:
```diff
@@ -333,6 +332,43 @@ func (s *v1AlphaAPIServer) getPodManifest(p *pod) (*schema.PodManifest, []byte,
+//
+// 2) All others
+//
+// Assume that we cna simply get the cgroup of the pod's init PID, and use that.
```
s/cna/can

yifan-gu commented on rkt/api_service.go:
```diff
@@ -333,6 +332,43 @@ func (s *v1AlphaAPIServer) getPodManifest(p *pod) (*schema.PodManifest, []byte,
+func getPodCgroup(p *pod, pid int) (string, error) {
+	// Note, this file should always exists if the pid is known since it is
+	// written before the pid file
+	data, err := p.readFile("subcgroup")
```
Can we comment how the content of the file looks like for example?

euank commented:
Addressed comments @yifan-gu

yifan-gu commented:
LGTM, one small nit.

euank commented:
The `subcgroup` file was added in #2655 by @iaguis, so I'd like to wait on a word from him to make sure this is a correct use of it.

jonboulle commented:
LGTM, though I'm no expert. Another pair of eyes, @lucab ?

euank commented:
This doesn't work with stage1-kvm since it does not register with machined even when it writes a `subcgroup` file. Presumably this is a bug with it. I also strongly suspect that means that the bug of waiting a significant time in the `api_service` applies for both kvm and fly pods. Given that unexpected issue, it seems like this should be bumped forwards a milestone, though it's unfortunate since the issue it fixes I think has more visible user impact than a wrong cgroup does (broken metrics vs potential for broken everything).

lucab commented:
@euank sorry to hear that. I'm still wrapping my head around cgroups, but I think the analysis at https://github.com/coreos/rkt/issues/3060#issuecomment-239409022 still applies here. Plus, I'm not sure what is the intended usage of that `subgcroup` file, outside of GC. And I'm quite sure that somebody else before attempted some unification across stage1 cgroup handling. @alban or @iaguis may have some insights here.

euank commented:
I got back to this and figured out the problem with kvm; it was erroniously writing the `subcgroup` file even though it neither uses nspawn nor registers with machined. Will be fixed in #3107, where I also added a few lines of documentation so we don't have to feel bad about using the `subcgroup` file here :)

I thought more about whether I wanted to add an annotation for "machinedregistration" to the stage1, but I think the stage1 communicating back up really is the way to go; duplicating the logic for whether registration can/should happen is fragile. I also added in a `podcgroup` file instead of the `subcgroup` file, but realized immediately that there's no point in renaming it if we just document the existing one and make it work, though I'm quite open to arguments there.

So I guess what I'm saying is I think this is actually generally reasonable as an approach, though I'd love to hear from @iaguis or @alban about better options or what's wrong here.

iaguis commented on rkt/api_service.go:
```diff
@@ -333,6 +332,57 @@ func (s *v1AlphaAPIServer) getPodManifest(p *pod) (*schema.PodManifest, []byte,
+//
+// Assume that we can simply get the cgroup of the pod's init PID, and use that.
+func getPodCgroup(p *pod, pid int) (string, error) {
+	// Note, this file should always exists if the pid is known since it is
```
s/exists/exist/

euank commented:
Rebased now that the other PR is merged; this is now pending second opinion/lgtm, squash, and merge.

In an ideal world, I'd do better testing of the interaction with cadvisor than I have, but I personally don't think we need to block merge/release on that.

iaguis commented on rkt/api_service.go:
```diff
@@ -333,6 +332,57 @@ func (s *v1AlphaAPIServer) getPodManifest(p *pod) (*schema.PodManifest, []byte,
+//
+// This one can be tricky because after machined registration, the cgroup might change.
+// For these flavors, the cgroup it will end up in after machined moves it is
+// written to a file named `podcgroup`, so we use that instead.
```
s/podcgroup/subcgroup/

iaguis commented:
LGTM after the nit and squashing.

euank commented:
Addressed and squashed.
